### PR TITLE
Assorted changes to speed up complex map rendering against PostGIS database (includes TWKB)

### PR DESCRIPTION
**Pull request by aaime**
@jnh5y I could use a review on the TWKB part. It's a bit different than the original, in particular:

- Made it non thread safe to avoid excessive object allocation during rendering (it was allocating various objects per geometry read and then throwing them away). The current implementation mirrors pretty closely the existing GeoTools WKBReader
- Removed dependency on protobuf, all that was needed were some 3 methods to read varints
- Un-commented all test methods and made them pass, removed the one about degenerated one point geometries as not even PostGIS ST_GeomFromTWKB would read them (while it was reading the other examples).

### CODE REVIEW

File: modules/library/main/src/main/java/org/geotools/filter/AttributeExpressionImpl.java [hunk 1]
```diff
@@ -224,33 +235,17 @@ public Object evaluate(Object obj) {
                                     + ") in object ("
                                     + obj
                                     + ")",
-                            e.get());
+                            lastException);
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/filter/AttributeExpressionImpl.java]: Did not know the method existed to start with. I see no use in GeoTools, one in GeoServer... would the suppressed exceptions be logged transparenly by log4j or java logging?

[the remark this answers is not in the document]
Comment (elahrvivaz) [reply modules/library/main/src/main/java/org/geotools/filter/AttributeExpressionImpl.java]: I believe so. Using the scala console, I get something like:
```
java.lang.Exception
	at $line3.$read$$iw$$iw$.<init>(<console>:10)
	at $line3.$read$$iw$$iw$.<clinit>(<console>)
	at $line3.$eval$.$print$lzycompute(<console>:7)
	at $line3.$eval$.$print(<console>:6)
	at $line3.$eval.$print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:784)
	at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:1039)
	at scala.tools.nsc.interpreter.IMain$WrappedRequest$$anonfun$loadAndRunReq$1.apply(IMain.scala:636)
	at scala.tools.nsc.interpreter.IMain$WrappedRequest$$anonfun$loadAndRunReq$1.apply(IMain.scala:635)
	at scala.reflect.internal.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
	at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:19)
	at scala.tools.nsc.interpreter.IMain$WrappedRequest.loadAndRunReq(IMain.scala:635)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:567)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:563)
	at scala.tools.nsc.interpreter.ILoop.reallyInterpret$1(ILoop.scala:802)
	at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:836)
	at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:694)
	at scala.tools.nsc.interpreter.ILoop.processLine(ILoop.scala:404)
	at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:424)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply$mcZ$sp(ILoop.scala:925)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:911)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:911)
	at scala.reflect.internal.util.ScalaClassLoader$.savingContextLoader(ScalaClassLoader.scala:97)
	at scala.tools.nsc.interpreter.ILoop.process(ILoop.scala:911)
	at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:74)
	at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:87)
	at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:98)
	at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:103)
	at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at scala_maven_executions.MainHelper.runMain(MainHelper.java:164)
	at scala_maven_executions.JavaMainCallerInProcess.runInternal(JavaMainCallerInProcess.java:94)
	at scala_maven_executions.JavaMainCallerInProcess.run(JavaMainCallerInProcess.java:54)
	at scala_maven_executions.JavaMainCallerSupport.run(JavaMainCallerSupport.java:110)
	at scala_maven.ScalaConsoleMojo.doExecute(ScalaConsoleMojo.java:98)
	at scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:574)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:309)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:194)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:107)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:955)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:290)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:194)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
	Suppressed: java.lang.RuntimeException
		at $line4.$read$$iw$$iw$.<init>(<console>:12)
		at $line4.$read$$iw$$iw$.<clinit>(<console>)
		at $line4.$eval$.$print$lzycompute(<console>:7)
		at $line4.$eval$.$print(<console>:6)
		at $line4.$eval.$print(<console>)
		... 60 more
	Suppressed: java.lang.RuntimeException
		at $line5.$read$$iw$$iw$.<init>(<console>:12)
		at $line5.$read$$iw$$iw$.<clinit>(<console>)
		at $line5.$eval$.$print$lzycompute(<console>:7)
		at $line5.$eval$.$print(<console>:6)
		at $line5.$eval.$print(<console>)
		... 60 more
```

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+                dims += 1;
+            }
+
+            if ((dimensions & 0x01) > 0) {
+                metadata.setHasM(true);
+                metadata.setMprecision((dimensions & 0xE0) >> 5);
+            }
+            if ((dimensions & 0x02) > 0) {
+                metadata.setHasZ(true);
+                metadata.setZprecision((dimensions & 0x1C) >> 2);
```
Comment (elahrvivaz) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java]: It looks like these two ifs could be combined with the ones that increment the `dims` variable a few lines above

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+import org.locationtech.jts.io.WKBWriter;
+
+public class TWKBAttributeIO {
+    TWKBReader wkbr;
+    ByteArrayInStream inStream = new ByteArrayInStream(new byte[0]);
+    GeometryFactory gf;
+
+    /**
+     * A variation of TWKBReader tha optimizes conversion of collapsed point expressed using
+     * LiteCoordinateSequece
```
Comment (jnh5y) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java]: Typo:)  Should be "LiteCoordinateSequence".  (Had to find something to prove that I read the PR.;))

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+import org.locationtech.jts.geom.LinearRing;
+import org.locationtech.jts.geom.MultiLineString;
+import org.locationtech.jts.geom.MultiPolygon;
+import org.locationtech.jts.geom.Point;
+import org.locationtech.jts.geom.Polygon;
+import org.locationtech.jts.io.ByteArrayInStream;
+import org.locationtech.jts.io.WKBWriter;
+
+public class TWKBAttributeIO {
+    TWKBReader wkbr;
```
Comment (jnh5y) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java]: Is it worth renaming the variables here to be twkbr or twkbReader?

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    public GeometryFactory getGeometryFactory() {
+        return gf;
+    }
+
+    /**
+     * This method will convert a Well Known Binary representation to a JTS Geometry object.
+     *
+     * @param wkb te wkb encoded byte array
```
Comment (jnh5y) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java]: wkb te wkb?  

Can you clarify this and update the WTB references below in this comment block?

File: modules/library/main/src/main/java/org/geotools/filter/AttributeExpressionImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/filter/AttributeExpressionImpl.java]: This PR did not actually change that behavior, it was only reporting the last exception before, but sure, I can make that change.

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+     * @return a JTS Geometry object that is equivalent to the WTB representation passed in by param
+     *     wkb
+     * @throws IOException if more than one geometry object was found in the WTB representation, or
+     *     if the parser could not parse the WKB representation.
+     */
+    private Geometry wkb2Geometry(byte[] wkbBytes) throws IOException {
+        if (wkbBytes == null) return null;
+        try {
+            inStream.setBytes(wkbBytes);
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java]: Which reminds me, this class is not meant to be usable outside, I'll make it package private

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+            case twkbMultiPoint:
+                return readMultiPoint(metadata);
+            case twkbMultiLineString:
+                return readMultiLineString(metadata);
+            case twkbMultiPolygon:
+                return readMultiPolygon(metadata);
+            case twkbGeometryCollection:
+                return readGeometryCollection(metadata);
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java]: Good point, but checked the WKB reader, and it throws an exception in that case, will mimic that behavior.

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+    }
+
+    private CoordinateSequence readCoordinateSequenceRing(int size, TWKBMetadata metadata)
+            throws IOException {
+        CoordinateSequence seq = readCoordinateSequence(size, metadata);
+        return CoordinateSequences.ensureValidRing(csFactory, seq);
+    }
+
+    private int zigzagDecode(int input) {
+        return (input >> 1) ^ (-(input & 1));
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java]: I inherited this bit of code as is from Jim, did not really touch it. But yeah, best to have them consistent, I'll modify to use the unsigned one.

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    }
+
+    private TWKBMetadata readMetadata() throws IOException {
+        int geometryTypeAndPrecision = dis.readByte();
+        int geometryType = geometryTypeAndPrecision & 0x0F;
+        int precision = zigzagDecode((geometryTypeAndPrecision & 0xF0) >> 4);
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java]: Same as above

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java (hunk 3 above)
Comment (aaime) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java]: :-D

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java (hunk 4 above)
Comment (aaime) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBAttributeIO.java]: Renamed to twkbReader, it's used only in a few places, the longer name does not seem to be a bother.

**Comment by aaime** [comment]
Windows build failure is an Appveyor issue to resolve github.com, spurious. Travis works fine, review reports addressed, merging.

**Pull request merged**